---
title: "M200A Evolutionary Biology, ABC lab"
author: "Hayley Stansell"
date: "November 29, 2015"
output: word_document
---



###Exercise 1

#####a) Assume that N can be any value b/w 100 and 100,000. Draw 1 million values from the prior dist. of N. Use a uniform distribution for N. Assume that N~U[100, 100000]
```{r}
getwd()
setwd("C:\\Users\\Hayley\\eeb201hayleystansell")
getwd()

#?runif
#try simulating the coin flip thing from class
#prior<- runif(1e6)
#D_sim<- rbinom(n=1e6, size=100, p=prior)
#plot(D_sim)
#This produces a "black block", x-axis going from 0 to 1e6, y-axis going from 0 to 100....
#?plot
#how is a number the prior distribution???? 
prior<- runif(n=1e6, min = 100, max = 100000)
#prior
#plot(prior) #looks correct in terms of axes
#
```

#####b) For each value of N, simulate a TMRCA. Use same approach as on last week's assignment. Note that you are using N (vs. 2N) because you are dealing with haploid samples (2 haploid samples from 2 individuals, one c-some per individ.)

```{r}
rate<- 1/(prior)
mostrecent<- rexp(1e6, rate)
#######DOES NOT NEED TO BE A LOOP!, R automatically takes each value of "prior" and plugs it in in order to the "mostrecent" vector
#######Follows that NEITHER DOES NEXT STEP
#mostrecent
length(mostrecent)
```

#####c) For each TMRCA, add a Poisson number of mutations w/ the appropriate mutation rate. rate = mu = 1e-8 per base pair per c-some....for 100kb is 1e-3.... *2? for 2 c-somes??

```{r}
ratemu<- 1e-3
snpnumber<- rpois(1e6, mostrecent*ratemu*2)
#snpnumber
max(snpnumber)
length(snpnumber)
```

#####d) Set parameters to "choose" which draws from the prior give data that approx. the OBSERVED number of SNPs accept all values of "N" that give b/w 45 and 55 SNPs

```{r}
sims<- cbind(prior, snpnumber)
simslower<- subset(sims, sims[,2]>45)
simsupper<- subset(simslower, simslower[,2]<55)
posterior<-simsupper[,1]
#posterior
max(posterior)
min(posterior)
hist(posterior)
density(posterior)
plot(density(posterior))
```

#####e) I should be done!


###EXERCISE TWO

#####Make a density plot of your prior and posterior distributions of N. plot them on the same axes, label which line corresponds to the prior, which corresponds to the posterior

```{r}
density(posterior)
plot(density(posterior), main= "Posterior vs. Prior distributions", xlab= "Population size (value of N)", col= "blue")
density(prior)
#?lines
# "lines() adds a line to an existing plot --- lines(namevector, col = color)
lines(density(prior), col= "red")
#?legend
legend(3e04,4e-06, # places a legend at a specified place 
       c("Posterior", "Prior"), # defines the text in the legend
       
       lty=c(1,1), # legend symbols (lines)
       
       lwd=c(2.5,2.5),col=c("blue", "red")) # gives lines the correct color and weight
```

###EXERCISE THREE

#####What is the median value of the posterior distribution of N?

```{r}
median(posterior)
```

###EXERCISE FOUR

#####Generate a 95% credible interval for the posterior dist. of N (like a confidence interval). Note, in this framework, there actually is a 95% chance of the true parameter value falling in this region. Hint - use "quantile" function. 

```{r}
#?quantile
quantile(posterior, c(0.025, 0.975))

```

The actual parameter value should, with a 95% chance, fall somewhere between these two values. This range represents 95% of the data output distribution, centered around the mean...so they are the most likely values.  

###EXERCISE FIVE

#####How does the posterior dist. differ from the prior distribution? A descriptive answer will suffice. 

The posterior distribution is essentially more informative than the prior distribution, as a product of the restrictions placed on its "acceptable" values - which are defined in the ABC simulation process. The range of likely (or even possible) values of N which satisfy the model in the posterior 1) are greatly reduced compared with the prior, and 2) have a more defined, uneven distribution across that range (approximating a normal distribution, more), which indicates that a certain range of N's is more likely to satifsy the model (compared with the prior, which has an even distribution across its range). 

###EXERCISE SIX

#####Repeat your ABC analysis, but change the prior dist. of N to be U~[1000,1e6]. What is the mean, median, and 95% credible interval for the posterior? How does this differ from what you computed in questions 3-4 for the original prior? What does this tell you about the effect fo the prior distribution in Bayesian stats?

```{r}
prior6<- runif(n=1e6, min = 1000, max = 1000000)
rate6<- 1/(prior6)
mostrecent6<- rexp(1e6, rate6)
length(mostrecent6)
ratemu6<- 1e-3
snpnumber6<- rpois(1e6, mostrecent6*ratemu6*2)
#snpnumber
max(snpnumber6)
length(snpnumber6)
sims6<- cbind(prior6, snpnumber6)
simslower6<- subset(sims6, sims6[,2]>45)
simsupper6<- subset(simslower6, simslower6[,2]<55)
posterior6<-simsupper6[,1]
#posterior
hist(posterior6)
#
#?plot
plot(density(posterior6), main= "Posterior vs. Prior distributions", xlab= "Population size (value of N)", col= "green")
density(prior6)

lines(density(prior6), col= "purple")

legend(6e05,3e-06, 
       c("Posterior6", "Prior6"), # puts text in the legend
       #try putting "top right" and it will set to right place
       
       lty=c(1,1), 
       
       lwd=c(2.5,2.5),col=c("green", "purple")) 

plot(density(posterior6), main= "Posterior vs. Prior distributions", xlab= "Population size (value of N)", col= "green")
lines(density(prior6), col="purple")
lines(density(prior), col ="red")
lines(density(posterior), col ="black")
#graph done to compare both simulations together.....not the most useful thing
```

For this simulation, the mean value of N is `r mean(posterior6)`, the median is `r median(posterior6)`, 
and the 95% credible interval is the range, 
`r quantile(posterior6, c(0.025, 0.975))` , meaning that there is a 95% percent chance that the actual value for N falls within this range. 
For the previous simulation, the mean was `r mean(posterior)` , the median was `r median(posterior)`, 
and the 95% credible interval was the range, 
`r quantile(posterior, c(0.025, 0.975))`

Increasing the prior distribution by an order of magnitude resulted in a posterior distribution that appears more refined and which hones in around a range of values that are perhaps more likely to be the "true N value". The range for the 95% credible interval is smaller for the second simulation. This implies that the posterior stemming from a larger range of N's in the prior improves the degree of likelihood or credibility for the posterior. 


###EXERCISE SEVEN

#####If you wanted a more precise estimate of the pop. size (assume there is no sex-biased demographic history so you can easily extrapolate the total pop. size from the Y c-some pop. size and vice versa), would you be better off:

#####a) sequencing a bigger region of the Y chromosome?
#####or
#####b) sequencing the same amount of the *autosomes*? 
#####why?

Generally, it seems that the larger the prior distribution, the smaller (or more dense) the posterior distribution. In order to increase the prior for this simulation, you could choose to sequence more of the Y-chromosome, effectively increasing the range of "N" used in the prior. This should result in a more precise estimate as the range for the 95% credible interval in the posterior should decrease.  




##Estimation of Population Split Times--------------------------------------------------------------------

###EXERCISE ONE

#####What is the expected additional time we have to wait until the two chromosomes coalesce *once* they make it back into the ancestral population?

Since this is a haploid group, the expected time to coalescence in the ancestral population is just N (vs. 2N) generations, where N is the population size (or number of y-chromosomes). If the ancestral population is of size N= 100,000, then the expected time to coalesce is 100,000 generations. 


###EXERCISE TWO

#####What is the expected time until the two chromosomes coalesce with each other? Write this formula using tsplit as split time. (write this as a formula, not a number estimate)

TimeToCoalescence = Ngenerations + tsplit
(basically, the time the populations were divergent, plus the time it took for the chromosomes to then coalesce in the ancestral population)


###EXERCISE THREE

-know that that ancestral pop size is N= 100000 (1e5) (haploid)
-sequenced a 100kb region (mu for region is 1e-3)
-do not know tsplit (time POPULATION split....not the same as TMRCA)
  -range of tsplit can be 50000 to 1e6 (1 million gen)
-do not know divided population size (need estimates)
  -this influences the Ngenerations? should be Ngenerations = 1e5


#####a) draw 1 million values from the prior distribution of tsplit. tsplit can be any value between 50,000 and 1,000,000, assume that tsplit~U[1000,1000000] .............this is incongruent with the first exercise, so assuming the range min is 50,000.

```{r}
prior3<-runif(n=1e6, min=50000, max=1e6)
```

#####b) For each value of tsplit, simulate a TMRCA (again, mimic the first exercise)

```{r}


rate3<- 1/(prior3)
tsplit3<- rexp(1e6, rate3)
length(tsplit3)

TMRCA<- tsplit3 + 1e5 #1e5 is the expected ancestral coalescent time
length(TMRCA)

```

#####c) For each TMRCA, add a Poisson number of mutations

```{r}
ratemu3<- 1e-3
snpnumber3<- rpois(1e6, TMRCA*ratemu3*2)

#snpnumber
max(snpnumber3)
min(snpnumber3)
length(snpnumber3)
#hist(snpnumber3)

```

#####d) Choose which draws from the prior generate data that approximate the observed number of SNPs (600) in the actual data. 
#####   -accept values of tsplit that give between 550 and 650 SNPs

```{r}

sims3<- cbind(prior3, snpnumber3)
simslower3<- subset(sims3, sims3[,2]>550)
simsupper3<- subset(simslower3, simslower3[,2]<650)
posterior3<-simsupper3[,1]
#posterior
max(posterior3)
min(posterior3)
#hist(posterior3)

```

#####e) It should be all done!


###EXERCISE FOUR

#####Make a density plot of both the prior and posterior distributions of tsplit. Plot on the same axes, and label each dist.

```{r}
density(posterior3)
plot(density(posterior3), main= "Posterior vs. Prior distributions", xlab= "T-split times", col= "blue")
density(prior3)
#?lines
# "lines() adds a line to an existing plot --- lines(namevector, col = color)
lines(density(prior3), col= "red")
#?legend
legend(3e05,5.5e-07, # legend location 
       c("Posterior", "Prior"), # text in the legend
       
       lty=c(1,1), # legend symbols (lines)
       
       lwd=c(2.5,2.5),col=c("blue", "red")) #lines the correct color and width
```

###EXERCISE FIVE

#####What is the median value of the posterior distribution of tsplit?

```{r}
median(posterior3)
```

The median value of the posterior is `r median(posterior3)` generations. 

###EXERCISE SIX

#####Generate a 95% credible interval for the posterior. 

The 95% credible interval for the posterior of tsplit is the range, `r quantile(posterior3, c(0.025, 0.975))` , meaning that there is a 95% percent chance that the actual value for the split time falls within this range. 

###EXERCISE SEVEN

#####How does the posterior dist. differ from the prior dist? The degree to which the posterior differs from the prior relates to the amt. of info. in the data.

As in the first exercise, the posterior distribution differs from the prior in that, while covering the same range of values, the density of values is not equally distributed across that range. Instead, there is more of a normal-distribution-esque curve to the density of values. The curve centers roughly around a median value. In this particular simulation, the degree of difference between the prior and the posterior is not particularly dramatic due to the amount of input information (which is similar to the amount used in the first round of exercise 1). 

###EXERCISE EIGHT

#####What if we did not know the true value of N? Repeat the ABC approach to estimate tsplit, but this time, rather than assume N=100,000, draw N from a distribution ~U[1000, 1000000]

```{r}
priorA<-runif(n=1e6, min=50000, max=1e6) #set A is for the split time
priorB<-runif(n=1e6, min=1000, max=1e6)  #set B is for the ancestral time


rateA<- 1/(priorA)
tsplitA<- rexp(1e6, rateA)
length(tsplitA)

rateB<- 1/(priorB)
tsplitB<- rexp(1e6, rateB)
length(tsplitB)

TMRCAfinal<- tsplitA + tsplitB #tsplitB is the expected ancestral coalescent time
length(TMRCAfinal)

#add in mutations
ratemu3<- 1e-3
snpnumberfinal<- rpois(1e6, TMRCAfinal*ratemu3*2)

#snpnumber
max(snpnumberfinal)
min(snpnumberfinal)
length(snpnumberfinal)
#hist(snpnumber3)

#focusing on prior and posterior of tsplitA (the split time)

simsf<- cbind(priorA, snpnumberfinal)
simslowerf<- subset(simsf, simsf[,2]>550)
simsupperf<- subset(simslowerf, simslowerf[,2]<650)
posteriorf<-simsupperf[,1]
#posterior
max(posteriorf)
min(posteriorf)
#hist(posteriorf)

```

###EXERCISE NINE

#####Do your estimates of the median and credible intervale of tsplit differ from the previous simulation?

The 95% credible interval for the posterior of tsplit in this simulation is the range, `r quantile(posteriorf, c(0.025, 0.975))` , and the median value for tsplit in this simulation is `r median(posteriorf)`

In the previous simulation, the 95% credible interval for the posterior of tsplit was `r quantile(posterior3, c(0.025, 0.975))`, and the median value for tsplit was `r median(posterior3)` . 
